### ISSUE TITLE
Question: Solving time & optimalizations

### ISSUE THREAD

**Issue by SonahtQ** [fabiospampinato/crypto-puzzle#3]
Does someone could potentially create solver in c/c++ and solve challanges faster than original js/ts implementation?

**Comment by fabiospampinato** [comment]
Oh yeah, by a huge amount if they are using some dedicated ASIC for the hashes, and if they are parallelizing the search etc. 

Potentially a different hash function could be used, I used xxHash in the past, which as far as I know nobody made a dedicated ASIC for yet. Still a native implementation could go much faster, and I didn't like the external third-party dependency, so I switched back to sha-512.

I should try to come up with an alternative algorithm where the search can't be parallelized, and where it's still ~free to generate on the server, then switching to something like xxHash can be much more appealing.

**Comment by fabiospampinato** [comment]
Just requiring some number of pbkdf2 rounds over the key should prevent some of the problem, no ASICs for that, which would encompass most of the work, and parallelization is still possible, but now much more expensive.

**Comment by fabiospampinato** [comment]
> Parallelization won't always be the case?

Some form of parallelization would still be possible with pbkdf2 rounds and/or xxhash, I can't say if it would "always" be possible though 🤔 It feels like if the client can't parallelize the search maybe it shouldn't be free for the server to generate the puzzle either, though the server is merely "just" generating the puzzle not, not solving, so there's maybe some way around that, right now I don't know.

Like maybe there's a way for the server to have some kind of secret which the client can't possibly guess which allows the server to generate puzzles for free, while it takes unparallelizable time for the client to solve. I just don't know an algorithm for that.

> Maybe there is some hash function from some web standard api. Then it should be as fast as it can be?

I'm already using functions that the browser provides, but the WebCrypto API wasn't really designed with performance in mind.

**Comment by fabiospampinato** [comment]
I should do some research to check if there isn't just a simpler known way to solve the parallelization problem differently. If there isn't making this library parallelized seems like a good idea.

**Comment by SonahtQ** [comment]
Node has Worker_Threads with similiar api to webworkers. But even without it (to support old node) lib could just spawn processes.

**Comment by fabiospampinato** [comment]
Yeah but I want the same library to work everywhere. I have worker_threads shimmed so that's not really a problem.

There's the question of "how many workers" though, like an attacker could always spawn more workers than you expected, like maybe one tunes the difficulty with the assumption that the client will be able to spawn up 4 workers, but the attacker could just spawn 400 or something, or parallelize across an indefinite number of machines potentially. If one assumes 400 workers, to make make it always somewhat expensive, then a legitimate user under an old mobile phone would have to waste a huge amount of time to solve these puzzles, so it doesn't seem ideal.

**Comment by fabiospampinato** [comment]
I feel like there's probably a much better theoretical solution here, I'm just not knowledgeable enough about this area at this time.

**Comment by fabiospampinato** [comment]
Yeah but how does the server know the solution without doing all that work? I just don't know the answer.

**Comment by SonahtQ** [comment]
Pre-creating puzzles could be one of options

**Comment by fabiospampinato** [comment]
But the server would still be doing the work though. That may be fine if the work in practice ends up being done during idle times or something, but it would be much better if it did approximately no work at all.

**Comment by fabiospampinato** [comment]
There seems to be a nice mathematical solution to this! I just need to understand it to implement it.

Some pointers for future me, or the curious:
- https://en.wikipedia.org/wiki/LCS35
- https://crypto.stackexchange.com/questions/9327/parallel-resistant-proof-of-work-scheme/9331#9331
- https://asecuritysite.com/encryption/pow2

**Comment by SonahtQ** [comment]
> There's the question of "how many workers" though, like an attacker could always spawn more workers than you expected, like maybe one tunes the difficulty with the assumption that the client will be able to spawn up 4 workers, but the attacker could just spawn 400 or something, or parallelize across an indefinite number of machines potentially.

Then that way, algorithm can be called "proof of resources&work" like not much people have that amount of cores/machines avaliable. It still usable. I'm thinking, the optimalization problem can be more important.

**Comment by pcfreak30** [comment]
I am doing research and am also interested in this problem for a web3/crypto/p2p project. Found some related projects, though I'm not sure how this code compares design-wise?

- https://github.com/tevador/equix
- https://git.sequentialread.com/forest/pow-captcha

**Comment by fabiospampinato** [comment]
V5 generates [time-lock puzzles](https://people.csail.mit.edu/rivest/pubs/RSW96.pdf), which are not parallelizable, and they are still cheap to generate.
